packages/messaging/messaging.ts

```diff
@@ -0,0 +1,171 @@
+  const { by = 'port' } = options ?? {};
+
+  if (by === 'sendMessage') {
+    return sendBySendMessage<T>(type, payload);
+  }
+  return sendByPort<T>(type, payload);
+};
+
+const createPortUtils = (sendMessage: (message: unknown) => unknown) => {
+  function response<M extends Message>(message: Omit<M, 'payload'>) {
+    try {
+      sendMessage(message);
+    } catch (error) {
+      console.error('Error in `sendResponse`', error);
+    }
+  }
+
+  function handleError<M extends Message>(message: M, error: unknown) {
+    if (error instanceof Error) {
+      response({ type: (message.type + ERROR_SUFFIX) as MessageType, response: error.toString() });
+    } else {
+      response({ type: (message.type + ERROR_SUFFIX) as MessageType, response: 'Unknown error' });
+    }
+  }
+
+  return {
+    response,
+    handleError,
+  };
+};
+
+function addHandlerByPort<T extends MessageType>(
+  type: T,
+  handleMessage: (payload: Payload<T>, sender?: chrome.runtime.MessageSender) => Promise<Response<T>> | Response<T>,
+) {
+  chrome.runtime.onConnect.addListener(port => {
+    const { response, handleError } = createPortUtils(message => port.postMessage(message));
+    port.onMessage.addListener(async (message: Message) => {
+      if (message.type !== type) {
+        return;
+      }
+      try {
+        response({ type: message.type, response: await handleMessage(message.payload as Payload<T>, port.sender) });
+      } catch (error) {
+        handleError(message, error);
+      }
+    });
+  });
+}
+
+function addHandlerBySendMessage<T extends MessageType>(
+  type: T,
+  handleMessage: (payload: Payload<T>, sender?: chrome.runtime.MessageSender) => Promise<Response<T>> | Response<T>,
+) {
+  chrome.runtime.onMessage.addListener((message: Message, sender, sendResponse) => {
+    const { response, handleError } = createPortUtils(sendResponse);
+    if (message.type !== type) {
+      return;
+    }
+    (async () => {
+      try {
+        response({ type: message.type, response: await handleMessage(message.payload as Payload<T>, sender) });
+      } catch (error) {
+        handleError(message, error);
+      }
+    })();
+    return true;
+  });
+}
+
+type AddMessageHandlerOptions = {
+  /** @default 'both' */
+  by?: 'port' | 'sendMessage' | 'both';
+};
+
+function on<T extends MessageType>(
+  type: T,
+  handleMessage: (payload: Payload<T>, sender?: chrome.runtime.MessageSender) => Promise<Response<T>> | Response<T>,
+  options?: AddMessageHandlerOptions,
+) {
+  const { by = 'both' } = options ?? {};
+  if (by === 'sendMessage' || by === 'both') {
+    addHandlerBySendMessage(type, handleMessage);
+  }
+  if (by === 'port' || by === 'both') {
+    addHandlerByPort(type, handleMessage);
+  }
+}
+
+export const messaging = {
+  sendToCurrentTab,
+  send,
+  sendByPort,
+  sendBySendMessage,
+  on,
+  onByPort: addHandlerByPort,
+  onBySendMessage: addHandlerBySendMessage,
+};
```

Comment:
Some documentation / jsdoc when / how to use? From background, from content-script send etc?